arangod/RocksDBEngine/RocksDBThrottle.cpp

```diff
@@ -99,89 +101,94 @@ thread_local sPriorityInfo gThreadPriority = {false, 0, 0};
 
 // rocksdb flushes and compactions start and stop within same thread, no
 // overlapping
-//  (OSX 10.12 requires a static initializer for thread_local ... time_point on
-//  mac does not have
-//   one in clang 9.0.0)
-thread_local uint8_t gFlushStart[sizeof(std::chrono::steady_clock::time_point)];
+thread_local std::chrono::steady_clock::time_point flushStart = std::chrono::steady_clock::time_point{};
 
-//
 // Setup the object, clearing variables, but do no real work
-//
 RocksDBThrottle::RocksDBThrottle()
     : _internalRocksDB(nullptr),
-      _threadRunning(false),
+      _throttleState(ThrottleState::NotStarted),
       _replaceIdx(2),
       _throttleBps(0),
       _firstThrottle(true) {
   memset(&_throttleData, 0, sizeof(_throttleData));
 }
 
-//
 // Shutdown the background thread only if it was ever started
-//
-RocksDBThrottle::~RocksDBThrottle() { StopThread(); }
+RocksDBThrottle::~RocksDBThrottle() { stopThread(); }
 
-//
 // Shutdown the background thread only if it was ever started
-//
-void RocksDBThrottle::StopThread() {
-  if (_threadRunning.load()) {
-    {
-      CONDITION_LOCKER(guard, _threadCondvar);
-
-      _threadRunning.store(false);
-      _threadCondvar.signal();
-    }  // lock
-
-    _threadFuture.wait();
-
-    {
-      CONDITION_LOCKER(guard, _threadCondvar);
-
-      _internalRocksDB = nullptr;
-      _delayToken.reset();
-    }  // lock
-  }    // if
-}  // RocksDBThrottle::StopThread
+void RocksDBThrottle::stopThread() {
+  ThrottleState state = _throttleState.load(std::memory_order_relaxed);
+
+  while (state != ThrottleState::Done) {
+    // we cannot shut down while we are currently starting
+    if (state == ThrottleState::NotStarted) {
+      // NotStarted => Done
+      if (_throttleState.compare_exchange_strong(state, ThrottleState::Done)) {
+        break;
+      }
+    } else if (state == ThrottleState::Running) {
+      // Started => ShuttingDown
+      if (_throttleState.compare_exchange_strong(state, ThrottleState::ShuttingDown)) {
+        {
+          CONDITION_LOCKER(guard, _threadCondvar);
+          _threadCondvar.signal();
+        }
+        _threadFuture.wait();
+
+        TRI_ASSERT(_throttleState.load() == ThrottleState::ShuttingDown);
+        _throttleState.store(ThrottleState::Done);
+
+        {
+          CONDITION_LOCKER(guard, _threadCondvar);
+
+          _internalRocksDB = nullptr;
+          _delayToken.reset();
+        }
+        break;
+      }
+    }
+
+    // wait until startup has finished
+    TRI_ASSERT(state == ThrottleState::Starting);
```

Comment:
I think this assert is not correct. Suppose we reach the `_throttleState.compare_exchange_strong` in line 127, but in the meantime `_throttleState` has been updated to Running. I think this could be fixed by removing the else in line 130, because the `compare_exchange` updates `state` (assuming stopThread cannot be called by multiple threads).